torchqdynamics/ode_adjoint_qsolver.py

```diff
@@ -0,0 +1,212 @@
+from __future__ import annotations
+
+from abc import abstractmethod
+
+import torch
+import torch.nn as nn
+from torch import Tensor
+from torch.autograd.function import FunctionCtx
+
+from .ode_forward_qsolver import ODEForwardQSolver
+from .progress_bar import tqdm
+from .solver_options import ODEAdaptiveStep, ODEFixedStep
+from .solver_utils import add_tuples, none_to_zeros_like
+
+
+class ODEAdjointQSolver(ODEForwardQSolver):
+    GRADIENT_ALG = ['autograd', 'adjoint']
+
+    def run(self):
+        super().run()
+
+        if self.gradient_alg == 'adjoint':
+            self._odeint_adjoint()
```

Comment:
The logic is kinda difficult to understand here. It would be better to copy the code again and to have
```python
if self.gradient_alg is None:
    self._odeint_inplace()
elif self.gradient_alg == 'autograd':
    self._odeint_main()
elif self.gradient_alg == 'adjoint':
    self._odeint_adjoint()
else:
    raise ValueError()
```
I added the error raising but maybe it's already somewhere else I'm not sure.

Reply:
Or alternatively to have
```python
if self.gradient_alg in [None, 'autograd']:
    super().run()
elif self.gradient_alg == 'adjoint':
    self._odeint_adjoint()
```